sdk/identity/Azure.Identity/src/ProcessRunner.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Identity
+{
+#pragma warning disable CA1001 // Types that own disposable fields should be disposable. Disposing of _process and _ctRegistration / _timeoutCtRegistration fields from outside may result in _tcs being incomplete or process handle leak.
+    internal sealed class ProcessRunner
+#pragma warning restore CA1001 // Types that own disposable fields should be disposable
+    {
+        private readonly IProcess _process;
+        private readonly TaskCompletionSource<string> _tcs;
+        private readonly CancellationToken _cancellationToken;
+        private readonly CancellationToken _timeoutCancellationToken;
+        private readonly CancellationTokenSource _timeoutCts;
+        private CancellationTokenRegistration _ctRegistration;
+        private CancellationTokenRegistration _timeoutCtRegistration;
+
+        public ProcessRunner(IProcess process, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            _process = process;
+            _cancellationToken = cancellationToken;
+            _tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            if (timeout.TotalMilliseconds >= 0)
+            {
+                _timeoutCts = new CancellationTokenSource();
+                _timeoutCancellationToken = _timeoutCts.Token;
```

Comment:
I still don't understand why do we need two token fields, they both seem to be used in exactly the same way

Reply:
`TrySetCanceled()` needs two `cancellationToken` instances to separate user-requested cancellation from timeout, e.g.:
```cs
try {
    var runner = new ProcessRunner(process, TimeSpan.FromSeconds(30), cancellationToken);
    return await runner.RunAsync().ConfigureAwait(false);
} catch (OperationCanceledException ex) when (ex.CancellationToken == cancellationToken) {
    // User-canceled
} catch (OperationCanceledException ex) {
    // Timeout
}
```

Reply:
But the code you've shown is outside of the ProcessRunner, it looks like implementation can use one linked CTS and a single registration.

Reply:
Yes, in that case linked CTS can be canceled by both user-provided CT and by timeout. With current implementation, it doesn't matter in which order user CT and timeout CT are checked. With linked CTS, we always have to check user CT first. This isn't obvious and can easily lead to a bug if that code is changed or refactored.

Reply:
Discussed offline